python/lsst/ctrl/oods/gen3ButlerBroker.py

```diff
@@ -0,0 +1,260 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import asyncio
+import collections
+import logging
+import os
+import shutil
+from lsst.ctrl.oods.butlerBroker import ButlerBroker
+from lsst.ctrl.oods.timeInterval import TimeInterval
+from lsst.ctrl.oods.imageData import ImageData
+from lsst.ctrl.oods.archiverName import ArchiverName
+from lsst.daf.butler import Butler
+from lsst.obs.base.ingest import RawIngestTask, RawIngestConfig
+from lsst.obs.base.utils import getInstrument
+from astropy.time import Time
+from astropy.time import TimeDelta
+import astropy.units as u
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Gen3ButlerBroker(ButlerBroker):
+    """Processes files on behalf of a Gen3 Butler.
+
+    Parameters
+    ----------
+    config: `dict`
+        configuration of this butler ingester
+    """
+    def __init__(self, config, publisher, publisher_queue):
+        self.archiver_name = ArchiverName().archiver_name
+        self.config = config
+        self.publisher = publisher
+        self.publisher_queue = publisher_queue
+
+        repo = self.config["repoDirectory"]
+        instrument = self.config["instrument"]
+        self.scanInterval = self.config["scanInterval"]
+        self.olderThan = self.config["filesOlderThan"]
+        self.collections = self.config["collections"]
+
+        self.repo_dir = self.config["repoDirectory"]
+        self.staging_dir = self.config["stagingDirectory"]
+        self.bad_file_dir = self.config["badFileDirectory"]
+
+        register = False
+        try:
+            butlerConfig = Butler.makeRepo(repo)
+            register = True
+        except FileExistsError:
+            butlerConfig = repo
+
+        instr = getInstrument(instrument)
+        run = instr.makeDefaultRawIngestRunName()
+        opts = dict(run=run, writeable=True, collections=self.collections)
+        self.butler = Butler(butlerConfig, **opts)
+
+        if register:
+            # Register an instrument.
+            instr.register(self.butler.registry)
+
+        cfg = RawIngestConfig()
+        cfg.transfer = "direct"
+        self.task = RawIngestTask(config=cfg,
+                                  butler=self.butler,
+                                  on_success=self.on_success,
+                                  on_ingest_failure=self.on_ingest_failure,
+                                  on_metadata_failure=self.on_metadata_failure)
+
+    def undef_metadata(self, filename):
+        """Return a sparsely initialized metadata dictionary
+        Parameters
```

Comment:
A blank line need to separate parameter description from the summary.

Reply:
Fixed